src/crh/mod.rs

```diff
@@ -50,18 +52,19 @@ pub trait TwoToOneCRH {
         + Default
         + CanonicalSerialize
         + CanonicalDeserialize;
-    type Parameters: Clone + Default;
+    type Parameters: Clone;
 
     fn setup<R: Rng>(r: &mut R) -> Result<Self::Parameters, Error>;
-    /// Evaluates this CRH on the left and right inputs.
-    ///
-    /// # Panics
-    ///
-    /// If `left_input.len() != Self::LEFT_INPUT_SIZE_BITS`, or if
-    /// `right_input.len() != Self::RIGHT_INPUT_SIZE_BITS`, then this method panics.
-    fn evaluate(
+
+    fn evaluate<T: Borrow<Self::Input>>(
+        parameters: &Self::Parameters,
+        left_input: T,
+        right_input: T,
+    ) -> Result<Self::Output, Error>;
+
+    fn compress<T: Borrow<Self::Output>>(
         parameters: &Self::Parameters,
-        left_input: &[u8],
-        right_input: &[u8],
+        left_input: T,
+        right_input: T,
```

Comment:
Why do we have separate `evaluate` and `compress` methods? If `evaluate` is meant to be used at the leaf node, then it can take just bytes, no?

Reply:
In `compress`, `T` is a borrow of `Self::Output` instead of `Self::Input`. In this implementation leaf hash do not need to be converted to byte. For example, if we use `Poseidon` as leaf hash (which takes field element and outputs field element) and `Poseidon` as inner hash, everything is in field element, and no bytes involved

Reply:
Right, so I guess I was wondering if we needed both `evaluate` and `compress`

Reply:
For now in merkle tree, `evaluate` is used only when converting leaf hash to inner hash. One alternative is to specify a function called `output_to_input`, but since `input` is not sized, we might need another associate type that represents `Borrow` of the `Input`, which adds complexity

Reply:
resolved via internal communication